server/router/src/main/java/org/infinispan/server/router/router/impl/hotrod/handlers/SniRouteHandler.java

```diff
@@ -0,0 +1,58 @@
+package org.infinispan.server.router.router.impl.hotrod.handlers;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInitializer;
+import io.netty.handler.ssl.SniHandler;
+import io.netty.handler.ssl.SslContext;
+import io.netty.util.DomainNameMapping;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.server.router.RoutingTable;
+import org.infinispan.server.router.logging.RouterLogger;
+import org.infinispan.server.router.routes.hotrod.NettyHandlerRouteDestination;
+import org.infinispan.server.router.routes.Route;
+import org.infinispan.server.router.routes.SniRouteSource;
+
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.Optional;
+
+public class SniRouteHandler extends SniHandler {
+
+    private static final RouterLogger logger = LogFactory.getLog(MethodHandles.lookup().lookupClass(), RouterLogger.class);
+
+    private final RoutingTable routingTable;
+
+    public SniRouteHandler(DomainNameMapping<? extends SslContext> mapping, RoutingTable routingTable) {
+        super(mapping);
+        this.routingTable = routingTable;
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        super.decode(ctx, in, out);
+
+        if(isHandShaked()) {
+            // At this point Netty has replaced SNIHandler (formally this) with SSLHandler in the pipeline.
+            // Now we need to add other handlers at the tail of the queue
+            logger.debug("Handshaked with hostname " + hostname());
+            Optional<Route<SniRouteSource, NettyHandlerRouteDestination>> route = routingTable.streamRoutes(SniRouteSource.class, NettyHandlerRouteDestination.class)
+                    .filter(r -> r.getRouteSource().getSniHostName().equals(this.hostname()))
+                    .findAny();
+
+            NettyHandlerRouteDestination routeDestination = route.orElseThrow(() -> logger.noRouteFound()).getRouteDesitnation();
+            ChannelInitializer<Channel> channelInitializer = routeDestination.getChannelInitializer();
+
+            ctx.pipeline().addLast(channelInitializer);
```

Comment:
No, we don't need to do that... 

This handlers extends Netty's `SniHandler` and the `SniHandler` removed itself from the pipeline after the handshake: https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SniHandler.java#L308

It's kinda ugly, since it tightly coupled with Netty's implementation but I couldn't find better solution.